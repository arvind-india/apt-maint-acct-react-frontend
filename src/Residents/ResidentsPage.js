import React from 'react'
import { Link } from 'react-router-dom'
import { connect } from 'react-redux'
import { Table } from 'reactstrap'
import { Router } from 'react-router-dom'

import {
  MdAdd,
  MdVisibility,
  MdEdit,
  MdDelete
} from 'react-icons/lib/md' // material design icons

import {
  Button
} from 'reactstrap'

import { history } from '../_helpers'
import { PrivateRoute, FlashMessage } from '../_components'
import { residentActions as actions, userActions } from '../_actions'
import { default as detailsPage } from './ResidentDetailsPage'

let url = '/residents'
let module = 'residents'

export class Residents extends React.Component {

  constructor(props) {
    super(props)
    this.handleDeleteModel = this.handleDeleteModel.bind(this)
  }

  componentDidMount() {
    //this.props.dispatch(actions.getAll())
    //this.props.dispatch(userActions.getAll())
    this.props.getAll()
    this.props.getAllUser()
  }

  render() {
    //console.log('Props in ResidentsPage: ', this.props)
    const { residents, alert, authzn, users, trackHistory } = this.props
    let models = residents
    let hist = trackHistory?history:{}
    return (
      <div>
        <h3>Residents List</h3>
        {alert.message && <FlashMessage text={alert.message} color={alert.color} delay={2100}/>}
        {models.loading && <em>Loading models...}</em>}
        {models.error && <span className="text-danger">ERROR: {models.error}</span>}
        {models.items && authzn && users.items && this.showList() }
        <Router history={hist}>
          <div>
            <PrivateRoute path={`${url}/:id`} component={detailsPage} />
          </div>
        </Router>
      </div>
    )
  }

/*
  handleDeleteModel(id) {
    console.log('Deleting Resident with id: ', id)
    //return (e) => this.props.dispatch(actions.delete(id))
    this.props.dispatch(actions.delete(id))
    this.props.dispatch(actions.getAll()) // get list after deletion of a model
  } */
  showList(){
    const { residents } = this.props
    let models = residents
    // make userId -> name array

    return <Table>
      <thead>{ this.headerRow() }</thead>
      <tbody>
        {models.items.map((model, index) => this.bodyRow(model, index))}
      </tbody>
    </Table>
  }
  headerRow() {
    return <tr>
      <th>#</th>
      <th>User Name</th>
      <th>First Name</th>
      <th>Last Name</th>
      <th>Type</th>
      <th>Occupied On</th>
      <th>Vacated On</th>
      <th>Actions {this.addLink()}</th>
    </tr>
  }
  bodyRow(model, index) {
    const { users } = this.props
    let userNames = []
    //console.log('users : ', users)
    users.items.forEach(each => userNames[each.id] = each.name)
    //console.log('userNames: ', userNames)

    return <tr key={model.id}>
      <th scope="row">{index+1}</th>
      <td>{userNames[model.owner_id]}</td>
      <td>{model.first_name}</td>
      <td>{model.last_name}</td>
      <td>{model.is_a}</td>
      <td>{model.occupied_on}</td>
      <td>{model.vacated_on}</td>
      { this.showActions(model) }
    </tr>
  }
  addLink() {
    const { authzn } = this.props
    return authzn.allowsAdd ?
      <Link to={{ pathname: `${url}/0`, state: this.newModel() }} title="Add"><MdAdd/></Link> :
      ''
  }

  newModel() {
    return {
      model: {
        id: 0,
        owner_id: 0,
        first_name:'',
        last_name: '',
        is_a: '',
        occupied_on: '',
        vacated_on: ''
      }
    }
  }

  showActions(model) {
    const { authzn } = this.props
    return <td>
            <Link
              to={{ pathname: `${url}/${model.id}`, state:{model: model} }}
              title={authzn.allowsEdit?"Edit":"View"}
            >{authzn.allowsEdit?<MdEdit/>:<MdVisibility/>}</Link>
            <Button
              color="link"
              title="Delete"
              onClick={() => this.handleDeleteModel(model.id)}
              hidden={!authzn.allowsDelete}
            ><MdDelete color="red"/></Button>
          </td>
  }

  handleDeleteModel(id) {
    if( window.confirm('Are you sure?') ) {
      console.log('Delete confirmed')
      //this.props.dispatch(actions.delete(id))
      //this.props.dispatch(actions.getAll()) // get list after deletion of a model
      this.props.delete(id)
      this.props.getAll() // get list after deletion of a model
    }
  }


}

function mapStateToProps(state) {
  const { residents, alert, authorizations, users } = state
  // const { user } = authentication
  const authzn = authorizations[module]
  const trackHistory = true  // added for unit testing; snapshot to be precise
  return {
//    user,
    residents,
    alert,
    authzn,
    users,
    trackHistory
  }
}

function mapDispatchToProps(dispatch) {
  return {
    getAll: () => {
      dispatch(actions.getAll())
    },
    getAllUser: () => {
      dispatch(userActions.getAll())
    },
    delete: (id) => {
      dispatch(actions.delete(id))
    }
  }
}

//const connectedResidentsPage = connect(mapStateToProps)(ResidentsPage)
//export { connectedResidentsPage as ResidentsPage }
export default connect(mapStateToProps, mapDispatchToProps)(Residents)
