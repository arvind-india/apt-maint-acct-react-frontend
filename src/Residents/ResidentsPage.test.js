// ResidentsPage.test.js

import React from 'react'
import { shallow } from 'enzyme'

import { Residents } from './ResidentsPage'

const mockProps = {
  trackHistory: false,  // for unit testing history is NOT required as it has 'keys' that are dynamic
  residents: {
    loading: false,
    error: false,
    items: [
      {
        id: 1,
        name: 'guest',
        firstName: 'guest',
        lastName: 'user',
        residentType: 'guest',
        occupiedOn: '2018-03-31',
        vacatedOn: '2020-03-31'
      },
      {
        id: 2,
        name: 'og1',
        firstName: 'g1',
        lastName: 'user',
        residentType: 'owner',
        occupiedOn: '2018-04-31',
        vacatedOn: '2020-04-31'
      },
      {
        id: 3,
        name: 'tg1',
        firstName: 't1',
        lastName: 'user',
        residentType: 'tenant',
        occupiedOn: '2018-05-31',
        vacatedOn: '2020-05-31'
      }
    ]
  },
  users: {
    items: [
      { name: 'guest' },
      { name: 'og1' },
      { name: 'tg1' }
    ]
  },
  alert: {message: "testingAlert"},
  authzn: {
    allowsAdd: true,
    allowsEdit: true,
    allowsDelete: true,
    condition: ''
  },
  getAll: jest.fn(),
  getAllUser: jest.fn(),
  delete: jest.fn()
}

const component = shallow(<Residents {...mockProps} />)

describe('ResidentsPage params test', () => {
  it('should render ResidentsPage without crashing', () => {
    component
  })
  it('should render initial layout', () => {
    expect(component.getElements()).toMatchSnapshot()
  })
})

describe('ResidentsPage rows test', () => {
  let numberOfMockResidents = mockProps.residents.items.length
  let tbody = component.find('tbody')
  let rows = tbody.find('tr')
  it('should have '+numberOfMockResidents+' rows', () => {
    expect(rows.length).toEqual(numberOfMockResidents)
  })
})
