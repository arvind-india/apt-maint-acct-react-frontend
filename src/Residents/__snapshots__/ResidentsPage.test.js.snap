// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResidentsPage params test should render initial layout 1`] = `
Array [
  <div>
    <h3>
        Residents List
    </h3>
    <FlashMessage
        color={undefined}
        delay={2100}
        text="testingAlert"
    />
    <Table
        responsiveTag="div"
        tag="table"
    >
        <thead>
            <tr>
                <th>
                    #
                </th>
                <th>
                    User Name
                </th>
                <th>
                    First Name
                </th>
                <th>
                    Last Name
                </th>
                <th>
                    Type
                </th>
                <th>
                    Occupied On
                </th>
                <th>
                    Vacated On
                </th>
                <th>
                    Actions 
                    <Link
                        replace={false}
                        title="Add"
                        to={
                            Object {
                                "pathname": "/residents/0",
                                "state": Object {
                                  "model": Object {
                                    "first_name": "",
                                    "id": 0,
                                    "is_a": "",
                                    "last_name": "",
                                    "occupied_on": "",
                                    "owner_id": 0,
                                    "vacated_on": "",
                                  },
                                },
                              }
                        }
                    >
                        <MdAdd />
                    </Link>
                </th>
            </tr>
        </thead>
        <tbody>
            <tr>
                <th
                    scope="row"
                >
                    1
                </th>
                <td>
                    tg1
                </td>
                <td />
                <td />
                <td />
                <td />
                <td />
                <td>
                    <Link
                        replace={false}
                        title="Edit"
                        to={
                            Object {
                                "pathname": "/residents/1",
                                "state": Object {
                                  "model": Object {
                                    "firstName": "guest",
                                    "id": 1,
                                    "lastName": "user",
                                    "name": "guest",
                                    "occupiedOn": "2018-03-31",
                                    "residentType": "guest",
                                    "vacatedOn": "2020-03-31",
                                  },
                                },
                              }
                        }
                    >
                        <MdEdit />
                    </Link>
                    <Button
                        color="link"
                        hidden={false}
                        onClick={[Function]}
                        tag="button"
                        title="Delete"
                    >
                        <MdDelete
                            color="red"
                        />
                    </Button>
                </td>
            </tr>
            <tr>
                <th
                    scope="row"
                >
                    2
                </th>
                <td>
                    tg1
                </td>
                <td />
                <td />
                <td />
                <td />
                <td />
                <td>
                    <Link
                        replace={false}
                        title="Edit"
                        to={
                            Object {
                                "pathname": "/residents/2",
                                "state": Object {
                                  "model": Object {
                                    "firstName": "g1",
                                    "id": 2,
                                    "lastName": "user",
                                    "name": "og1",
                                    "occupiedOn": "2018-04-31",
                                    "residentType": "owner",
                                    "vacatedOn": "2020-04-31",
                                  },
                                },
                              }
                        }
                    >
                        <MdEdit />
                    </Link>
                    <Button
                        color="link"
                        hidden={false}
                        onClick={[Function]}
                        tag="button"
                        title="Delete"
                    >
                        <MdDelete
                            color="red"
                        />
                    </Button>
                </td>
            </tr>
            <tr>
                <th
                    scope="row"
                >
                    3
                </th>
                <td>
                    tg1
                </td>
                <td />
                <td />
                <td />
                <td />
                <td />
                <td>
                    <Link
                        replace={false}
                        title="Edit"
                        to={
                            Object {
                                "pathname": "/residents/3",
                                "state": Object {
                                  "model": Object {
                                    "firstName": "t1",
                                    "id": 3,
                                    "lastName": "user",
                                    "name": "tg1",
                                    "occupiedOn": "2018-05-31",
                                    "residentType": "tenant",
                                    "vacatedOn": "2020-05-31",
                                  },
                                },
                              }
                        }
                    >
                        <MdEdit />
                    </Link>
                    <Button
                        color="link"
                        hidden={false}
                        onClick={[Function]}
                        tag="button"
                        title="Delete"
                    >
                        <MdDelete
                            color="red"
                        />
                    </Button>
                </td>
            </tr>
        </tbody>
    </Table>
    <Router
        history={Object {}}
    >
        <div>
            <PrivateRoute
                component={[Function]}
                path="/residents/:id"
            />
        </div>
    </Router>
</div>,
]
`;
